print verbose error if ofi provider not available

carns opened the pull request:
If initial `fi_getinfo()` fails, then query libfabric to determine if the desired provider is available.  If not, then display a more helpful error message.

I'm marking this PR as a draft, because the resulting error message is somewhat cluttered by the additional context (function, line number, etc.) prepended by the logging functions:

```
# NA -- Error -- /home/carns/working/src/mochi/mercury/src/na/na_ofi.c:1513
 # na_ofi_getinfo(): Requested OFI provider "verbs;ofi_rxm" (derived from
  "verbs" Mercury protocol) is not available.
# NA -- Error -- /home/carns/working/src/mochi/mercury/src/na/na_ofi.c:1517
 # na_ofi_getinfo(): Please re-compile libfabric with support for
  "verbs;ofi_rxm" or use one of the following available providers:
  tcp;ofi_rxm,tcp,sockets
# NA -- Error -- /home/carns/working/src/mochi/mercury/src/na/na_ofi.c:1526
 # na_ofi_getinfo(): fi_getinfo() failed, rc: -61 (No data available)
...
```

Is there a logging variant that doesn't do that, or should we make one?  Or just leave it as is?

carns commented:
Fixes #395

soumagne commented:
Thanks a lot Phil! I think it looks good! Just a few minor comments.

soumagne reviewed (approved):
Also if you look there: https://travis-ci.com/github/mercury-hpc/mercury/jobs/386470772
You'll see that this test is failing and it will tell you how to reformat it. You can run that same command from the top source directory:
```
 ./Testing/script/travis_format.sh src/na/na_ofi.c
```
or to reformat it automatically you can do:
```
clang-format --style=file src/na/na_ofi.c
```

carns commented on src/na/na_ofi.c:
```diff
@@ -1455,6 +1455,74 @@ na_ofi_getinfo(enum na_ofi_prov_type prov_type, struct fi_info **providers)
         hints,                      /* In: Hints to filter providers */
         providers);                 /* Out: List of matching providers */
+    if(rc != 0) {
+        /* getinfo failed.  This could be because Mercury was
```
fixed and force pushed

carns commented:
Ok, I refactored to a separate function and corrected the formatting (thanks for the tip!).

Thoughts on the error message format?  Just go with it as is?

soumagne commented:
Great thanks! I think the error message looks reasonable to me, we can still tweak it more later if needed.

soumagne commented:
Just a few more warnings to fix for you :) see https://mercury-cdash.hdfgroup.org/viewBuildError.php?type=1&buildid=851
```
/home/travis/build/mercury-hpc/mercury/src/na/na_ofi.c: In function 'na_ofi_provider_check':
[CTest: warning matched] /home/travis/build/mercury-hpc/mercuryna/na_ofi.c:1415:23: warning: conversion to 'size_t' {aka 'long unsigned int'} from 'int' may change the sign of the result [-Wsign-conversion]
 1415 |             avail_len += strlen(cur->fabric_attr->prov_name) + 1;
      |                       ^~
[CTest: warning matched] /home/travis/build/mercury-hpc/mercuryna/na_ofi.c:1415:26: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
 1415 |             avail_len += strlen(cur->fabric_attr->prov_name) + 1;
      |                          ^~~~~~
[CTest: warning matched] /home/travis/build/mercury-hpc/mercuryna/na_ofi.c:1421:30: warning: conversion to 'size_t' {aka 'long unsigned int'} from 'int' may change the sign of the result [-Wsign-conversion]
 1421 |     avail = calloc(avail_len + 1, 1);
      |                    ~~~~~~~~~~^~~
```

soumagne commented:
You can see those warnings by going to mercury-cdash.hdfgroup.org and click on the number in the "Warn" column.

carns commented:
Oh man, that's tough!  I think I have it fixed now :)

soumagne commented:
;) right I know I have the whole scrutinizer set up

soumagne commented:
Alright it's merged now :)
